### DISCUSSION TITLE
How to move body during ContactAdded?

### DISCUSSION THREAD

**Discussion by LittleCodingFox** [jrouwe/JoltPhysics#973] [Q&A] [answered]
I'm trying to make a teleporter and I want the character to be teleported when touching a teleport surface, but the app freezes instead. I thought it was due to locks, but even adding lock checking code to use the `NoLock` variants of `BodyInterface`, it will then assert that it has no rights (`BodyAccess::sCheckRights(BodyAccess::sPositionAccess, BodyAccess::EAccess::ReadWrite)`)

I looked at the issues and discussions and I'm not having much luck figuring this out, and I'd really appreciate some help!

Thank you very much!

**Comment by jrouwe** [comment, marked as answer]
You can't move a body in the contact callback, you need to remember that the contact happened and do it after the physics simulation step finishes (and possibly reject the contact point so that the other body is not affected).

**Reply by LittleCodingFox** [reply]
Oh okay, in that case I can just a main thread dispatch and handle the logic there then!

I imagine I should do this for all callbacks except ContactValidate?

**Reply by MrOnlineCoder** [reply]
I also wanted to find this out - does this rule apply only to body movement or to basically any modification to the system's state, that are not allowed during execution of any callbacks - like deletion of the body (bullet reaches it's target on collision)?

**Reply by jrouwe** [reply]
At the time that the contact validate/added functions are called, many threads are working on the bodies at the same time. So it is not safe to do anything but to read the state of the two involved bodies to return `ValidateResult` in case of `OnContactValidate` or set the state of `ContactSettings` in case of `OnContactAdded/Persisted`. In `OnContactRemoved` the involved bodies may not even exist anymore, so you can only record that the contact was lost.
